os/20201103-full-disk-raid.md

```diff
@@ -0,0 +1,208 @@
+
+Dracut glue detects these config stanzas, saves the corresponding on-disk data to RAM before the Ignition disks phase, and restores it after disks phase completes.  The glue also replicates the boot sector when replicating BIOS-BOOT.
+
+Users should only create replicated volumes using the official FCC sugar.  However, each piece of the replication logic (root partition, boot partition, ESP, and BIOS-BOOT + MBR) functions independently, and an advanced user can create nonstandard partition configurations (for example, RAID-1 boot with RAID-5 root) by manually configuring partitions in their FCC.
+
+## FCC sugar
+
+```yaml
+variant: fcos
+version: 1.3.0
+storage:
+  boot_raid:
+    devices:
+      # Two or more devices are allowed
+      - /dev/vda
+      - /dev/vdb
+    # Optional, defaults to x86_64
+    layout: x86_64
+    # Optional, defaults to 0 (entire disk)
+    # Do we want this?
+    root_size_mib: 8192
+    # Optional, defaults to xfs
+    # Do we want this?
+    root_format: xfs
+```
+
+The `layout` option allows choosing from multiple disk layout templates hardcoded in FCCT.  We might use this to support CPU architectures with different partition requirements (e.g. ppc64le) or to change the disk layout in future OS releases.
+
+The sugar does not support configuring RAID spares, filesystem options, or mount options.
+
+## Rendered Ignition config
+
+```json
+{
+  "ignition": {
+    "version": "3.2.0"
+  },
+  "storage": {
+    "disks": [
+      {
+        "device": "/dev/vda",
+        "partitions": [
+          {
+            "label": "bios-1",
+            "sizeMiB": 1,
+            "typeGuid": "21686148-6449-6E6F-744E-656564454649"
+          },
+          {
+            "label": "esp-1",
+            "sizeMiB": 127,
+            "typeGuid": "C12A7328-F81F-11D2-BA4B-00A0C93EC93B"
+          },
+          {
+            "label": "boot-1",
+            "sizeMiB": 384
+          },
+          {
+            "label": "root-1",
+            "sizeMiB": 8192
+          }
+        ],
+        "wipeTable": true
+      },
+      {
+        "device": "/dev/vdb",
+        "partitions": [
+          {
+            "label": "bios-2",
+            "sizeMiB": 1,
+            "typeGuid": "21686148-6449-6E6F-744E-656564454649"
+          },
+          {
+            "label": "esp-2",
+            "sizeMiB": 127,
+            "typeGuid": "C12A7328-F81F-11D2-BA4B-00A0C93EC93B"
+          },
+          {
+            "label": "boot-2",
+            "sizeMiB": 384
+          },
+          {
+            "label": "root-2",
+            "sizeMiB": 8192
+          }
+        ],
+        "wipeTable": true
+      }
+    ],
+    "filesystems": [
+      {
+        "device": "/dev/disk/by-partlabel/esp-1",
```

Comment:
by-partlabel will not work on s390x : https://github.com/coreos/ignition-dracut/pull/100#issuecomment-529948403

The GPT partitioning scheme uses little endian by default whereas s390x is big endian causing the by-partlabel string to be interpreted incorrectly.

Reply:
Okay, we can have FCCT set partition UUIDs and use `/dev/disk/by-partuuid` instead.

In the long run, is the s390x partlabel issue being fixed?

Reply:
not sure if the issue is being actively tracked. i will talk to some IBM folks to see if they are aware of anything.

Reply:
Since s390x is being declared out of scope for now, we're going to stick with partition labels.  s390x, if added, will use a separate config template which can use partition UUIDs if still needed.